openstreetmap-carto.lua

```diff
@@ -0,0 +1,518 @@
+-- @param tags OSM tags
+-- @return z_order if an object with z_order, otherwise nil
+function z_order(tags)
+    local z = 0
+    for k, v in pairs(tags) do
+        if roads_info[k] and roads_info[k][v] then
+            z = math.max(z, roads_info[k][v].z)
+        end
+    end
+    return z ~= 0 and z or nil
+end
+
+--- Gets the roads table status for a set of tags
+-- @param tags OSM tags
+-- @return 1 if it belongs in the roads table, 0 otherwise
+function roads(tags)
+    for k, v in pairs(tags) do
+        if roads_info[k] and roads_info[k][v] and roads_info[k][v].roads then
+            if not (k ~= 'railway' or tags.service) then
+                return 1
+            elseif not excluded_railway_service[tags.service] then
+                return 1
+            end
+        end
+    end
+    return 0
+end
+
+-- Filtering on nodes, ways, and relations
+function filter_tags_generic(tags, n)
+    -- Delete tags listed in delete_tags
+    for tag, _ in pairs (tags) do
+        for _, d in ipairs(delete_tags) do
+            if tag == d then
+              tags[tag] = nil
+              break -- Skip this tag from further checks since it's deleted
+            end
+        end
+    end
+    -- By using a second loop for wildcards we avoid checking already deleted tags
+    for tag, _ in pairs (tags) do
+        for _, d in ipairs(delete_wildcards) do
+            if string.find(tag, d) then
+                tags[tag] = nil
+                break
+            end
+        end
+    end
+
+   -- Filter out objects that have no tags after deleting
+    if next(tags) == nil then
+        return 1, {}
+    end
+
+-- Convert layer to an integer
+    tags['layer'] = layer(tags['layer'])
+    return 0, tags
+end
+
+-- Filtering on nodes
+function filter_tags_node (keyvalues, numberofkeys)
+    return filter_tags_generic(keyvalues, numberofkeys)
+end
+
+-- Filtering on relations
+function filter_basic_tags_rel (keyvalues, numberofkeys)
+    -- Filter out objects that are filtered out by filter_tags_generic
+    local filter, keyvalues = filter_tags_generic(keyvalues, numberofkeys)
+    if filter == 1 then
+        return 1, keyvalues
+    end
+
+    -- Filter out all relations except route, multipolygon and boundary relations
+    if ((keyvalues["type"] ~= "route") and (keyvalues["type"] ~= "multipolygon") and (keyvalues["type"] ~= "boundary")) then
+        return 1, keyvalues
+    end
+
+    return 0, keyvalues
+end
+
+-- Filtering on ways
+function filter_tags_way (keyvalues, numberofkeys)
+    local filter = 0  -- Will object be filtered out?
+    local polygon = 0 -- Will object be treated as polygon?
+
+    -- Filter out objects that are filtered out by filter_tags_generic
+    filter, keyvalues = filter_tags_generic(keyvalues, numberofkeys)
+    if filter == 1 then
+        return filter, keyvalues, polygon, roads
+    end
+
+    polygon = isarea(keyvalues)
+
+    -- Add z_order column
+    keyvalues.z_order = z_order(keyvalues)
```

Comment:
This came over from CT where there's a clear difference between tags (input, which may include annoying things like addr:foo) and columns (output, where I can make sure the names never collide with Lua or postgres syntax), but I agree, in this that distinction isn't regularly made, so it should be `keyvalues['z_order']`